### PR TITLE
feat: adds CAB rules classes

### PR DESCRIPTION
**Pull request by lsirac**
Adds CAB rule classes (see go/cab-client)

### CODE REVIEW

File: oauth2_http/javatests/com/google/auth/oauth2/CredentialAccessBoundaryTest.java [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            .build();
+
+    CredentialAccessBoundary credentialAccessBoundary =
+        CredentialAccessBoundary.newBuilder()
+            .setRules(Arrays.asList(firstRule, secondRule))
+            .build();
+
+    assertEquals(2, credentialAccessBoundary.getAccessBoundaryRules().size());
+    assertEquals(firstRule, credentialAccessBoundary.getAccessBoundaryRules().get(0));
+    assertEquals(secondRule, credentialAccessBoundary.getAccessBoundaryRules().get(1));
```
Comment (TimurSadykov) [inline oauth2_http/javatests/com/google/auth/oauth2/CredentialAccessBoundaryTest.java:73]: assert AvailabilityCondition is set in rule1 and not set in rule2?

File: oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+     */
+    public static final class AvailabilityCondition {
+      private final String expression;
+
+      @Nullable private final String title;
+      @Nullable private final String description;
+
+      AvailabilityCondition(
+          String expression, @Nullable String title, @Nullable String description) {
+        this.expression = checkNotNull(expression);
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java:245]: we don't want blank string here right? Maybe checkArgument(!StringUtils.isBlank(expression))?

File: oauth2_http/javatests/com/google/auth/oauth2/CredentialAccessBoundaryTest.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+    assertEquals("expression", availabilityCondition.getExpression());
+    assertNull(availabilityCondition.getTitle());
+    assertNull(availabilityCondition.getDescription());
+  }
+
+  @Test
+  public void availabilityCondition_nullExpression_throws() {
+    try {
+      AvailabilityCondition.newBuilder().setExpression(null).build();
```
Comment (TimurSadykov) [inline oauth2_http/javatests/com/google/auth/oauth2/CredentialAccessBoundaryTest.java:250]: test for blank? (empty, whitespace)

File: oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+      }
+
+      public AccessBoundaryRule build() {
+        return new AccessBoundaryRule(
+            availableResource, availablePermissions, availabilityCondition);
+      }
+    }
+
+    /**
+     * An optional condition that can be used as part of a {@link CredentialAccessBoundary} to
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java:234]: Some details on what kind of expression it is, how it is used?

File: oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+      public static final class Builder {
+        private String expression;
+
+        @Nullable private String title;
+        @Nullable private String description;
+
+        private Builder() {}
+
+        /** */
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java:276]: remove empty?

File: oauth2_http/javatests/com/google/auth/oauth2/CredentialAccessBoundaryTest.java [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+      fail("Should fail.");
+    } catch (IllegalArgumentException e) {
+      assertEquals("The provided list has more than 10 access boundary rules.", e.getMessage());
+    }
+  }
+
+  @Test
+  public void credentialAccessBoundary_toJson() {
+    AvailabilityCondition availabilityCondition =
+        AvailabilityCondition.newBuilder().setExpression("expression").build();
```
Comment (TimurSadykov) [inline oauth2_http/javatests/com/google/auth/oauth2/CredentialAccessBoundaryTest.java:110]: Let's add title and description to make sure they make their way to json

File: oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+      accessBoundaryRules.add(checkNotNull(rule));
+      return this;
+    }
+
+    public CredentialAccessBoundary build() {
+      return new CredentialAccessBoundary(accessBoundaryRules);
+    }
+  }
+
+  /** Defines an upper bound of permissions on a particular resource. */
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java:134]: elaborate with small example?

**Review by TimurSadykov — CHANGES_REQUESTED** [review]

File: oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+      public Builder setAvailablePermissions(List<String> availablePermissions) {
+        this.availablePermissions = new ArrayList<>(checkNotNull(availablePermissions));
+        return this;
+      }
+
+      /**
+       * Adds a permission that can be used on the resource. This should be an IAM role prefixed by
+       * inRole.
+       *
+       * <p>e.g. "inRole:roles/storage.objectViewer".
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java:208]: nit: e.g. --> for example, per google tech writing guidelines

File: oauth2_http/javatests/com/google/auth/oauth2/CredentialAccessBoundaryTest.java [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import static org.junit.Assert.*;
```
Comment (elharo) [inline oauth2_http/javatests/com/google/auth/oauth2/CredentialAccessBoundaryTest.java:34]: google style does not allow wildcard imports

File: oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java [hunk 10]
```diff
@@ -167,7 +167,10 @@ public CredentialAccessBoundary build() {
           "The provided availableResource is either null or empty.");
       this.availableResource = availableResource;
 
-      this.availablePermissions = new ArrayList<>(checkNotNull(availablePermissions));
+      checkArgument(
+          availablePermissions != null && !availablePermissions.isEmpty(),
+          "The list of provided availablePermissions is either null or empty.");
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java:173]: These should be two checks with two different messages. In general avoid messages like "Thing X is wrong or Thing Y is wrong" in favor of simply "Thing X is wrong"

File: oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import static com.google.common.base.Preconditions.*;
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java:34]: google style doesn't use wildcard imports

File: oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+ * information.</a>
+ */
+public final class CredentialAccessBoundary {
+
+  private static final int RULES_SIZE_LIMIT = 10;
+
+  private final List<AccessBoundaryRule> accessBoundaryRules;
+
+  CredentialAccessBoundary(List<AccessBoundaryRule> accessBoundaryRules) {
+    checkArgument(accessBoundaryRules != null, "The provided list of accessBoundaryRules is null.");
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java:55]: Preconditions.checkNotNull

**Review by elharo — APPROVED** [review]

**Review by TimurSadykov — APPROVED** [review]
LGTM

**Pull request merged**